MAINT: Change the type of error in numpy.ndarray.fill

kianasun opened the pull request:
I think it is a ``TypeError`` when the value is not a scalar.
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

eric-wieser commented:
A change like this needs a test, if only so we can see when the behavior changes

kianasun committed:
MAINT: Change the type of error in numpy.ndarray.fill

kianasun commented:
Hi @eric-wieser , I added tests for it. I also updated the document because it seems that `value` does not need to be a scalar value when the `dtype` of `a` is `"O"`.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -3003,12 +3003,13 @@
     value : scalar
-        All elements of `a` will be assigned this value.
+        All elements of `a` will be assigned this value. If `a` is not a
+        (Python) object, `value` must be a scalar value.
```
How about "`value` must be a scalar unless `a` is an object array?

eric-wieser commented:
I'm actually not sure TypeError _is_ appropriate here. I can pass in `np.array(0)` and that has the same type as `np.array([0])` which fails. So I'm inclined to close this. Thoughts, @seberg?

What might be cool is to add a new `class ShapeError(ValueError)` exception to numpy, which we can use in exactly this type of situation.

seberg commented:
I would agree that disallowing 0D arrays completely would make sense (and they are pretty rare). That would require a slightly annoying deprecation though. As long as we let them through, I cannot say I care too much, but it probably is not be worth the trouble, than to be honest.

If ShapeError also encompasses BroadcastError's, it could make sense, I guess.

mattip commented:
Closing. If someone feels strongly about it they can reopen with an updated PR
